contrib/win32/win32compat/fileio.c

```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
 		goto cleanup;
 	}
 
-	/* read the link data from the passed symlink */
-	resolved_link = fileio_readlink_internal(wpath);
-	if (resolved_link == NULL) {
-		/* errorno will have been set by called function */
+	/* obtain a handle to send to deviceioctl  */
+	handle = CreateFileW(wpath, 0, 0, NULL, OPEN_EXISTING, 
+		FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT, 0);
+	if (handle == INVALID_HANDLE_VALUE) {
+		errno = errno_from_Win32LastError();
+		goto cleanup;
+	}
+
+	/* send a request to the file system to get the real path */
+	reparse_buffer = (PREPARSE_DATA_BUFFER_SYMLINK) malloc(MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+	DWORD dwBytesReturned = 0;
+	if (DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, NULL, 0,
+		(LPVOID) reparse_buffer, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &dwBytesReturned, 0) == 0) {
+		errno = errno_from_Win32LastError();
 		goto cleanup;
 	}
 
-	if ((output = utf16_to_utf8(resolved_link)) == NULL) {
-		errno = ENOMEM;
+	/* ensure file is actually symlink */
+	if (reparse_buffer->ReparseTag != IO_REPARSE_TAG_SYMLINK) {
+		errno = EINVAL;
+		goto cleanup;
+	}
+
+	/* the symlink structure has a 'Print Name' value that is displayed to the
+	 * user which is different from the actual value it uses for redirection
+	 * called the 'Substitute Name'; since the Substitute Name has an odd format
+	 * that begins with \??\ and it appears that CreateSymbolicLink() always
+	 * formats the PrintName value consistently we will just use that
+	 */
+	int symlink_nonnull_size = reparse_buffer->PrintNameLength;
+	wchar_t * symlink_nonnull = &reparse_buffer->PathBuffer[reparse_buffer->PrintNameOffset / sizeof(WCHAR)];
+
+	/* allocate area to hold a null terminated version of the string */
+	if ((linkpath = malloc(symlink_nonnull_size + sizeof(wchar_t))) == NULL) {
 		goto cleanup;
 	}
 
-	/* ensure output buffer is large enough forward slash and the string */
-	ssize_t out_size = strlen(output);
-	if (1 + out_size > bufsiz) {
-		errno = ENAMETOOLONG;
+	/* copy the data out of the reparse buffer and add null terminator */
```

Comment:
symlink_nonnull can be a relative path as well?
If yes, then we are missing the logic to deal with relative paths.

Reply:
Yes, this was actually a major change for readlink() since the previous version I submitted.  I tested readlink() on Linux and it  just returns the symbolic link data for relative links; it does not attempt to construct the absolute path based on the link location and the link's parent.  So, for purposes of consistency, I made it act like Linux behaves.  The man pages actually say that readlink() will "place the contents of the symbolic link referred to by path in the buffer buf".  So that's what's it's doing now with the exception of the absolute path which it normalizes to have the forward slash.

Reply:
Ok.